allennlp/modules/span_extractors/bidirectional_endpoint_span_extractor.py

```diff
@@ -0,0 +1,230 @@
+            return forward_combined_dim + backward_combined_dim + \
+                   self._span_width_embedding.get_output_dim()
+        return forward_combined_dim + backward_combined_dim
+
+    @overrides
+    def forward(self, # pylint: disable=arguments-differ
+                sequence_tensor: torch.FloatTensor,
+                indices: torch.LongTensor,
+                sequence_mask: torch.LongTensor = None,
+                indices_mask: torch.LongTensor = None) -> torch.FloatTensor:
+
+        # Both of shape (batch_size, sequence_length, embedding_size / 2)
+        forward_sequence, backward_sequence = torch.split(sequence_tensor, int(self._input_dim / 2), -1)
+        forward_sequence = forward_sequence.contiguous()
+        backward_sequence = backward_sequence.contiguous()
+
+        # shape (batch_size, num_spans)
+        span_starts, span_ends = [index.squeeze(-1) for index in indices.split(1, dim=-1)]
+
+        if indices_mask is not None:
+            span_starts *= indices_mask
+            span_ends *= indices_mask
+
+        # We want `exclusive` span starts, so we remove 1 from the forward span starts
+        # as the AllenNLP ``SpanField`` is inclusive.
+        # shape (batch_size, num_spans)
+        exclusive_span_starts = span_starts - 1
+        # shape (batch_size, num_spans)
+        start_sentinel_mask = (exclusive_span_starts == -1).float().unsqueeze(-1)
+
+        # We want `exclusive` span ends for the backward direction
+        # (so that the `start` of the span in that direction is exlusive), so
+        # we add 1 to the span ends as the AllenNLP ``SpanField`` is inclusive.
+        exclusive_span_ends = span_ends + 1
+
+        if sequence_mask is not None:
+            # shape (batch_size)
+            sequence_lengths = get_lengths_from_binary_sequence_mask(sequence_mask)
+        else:
+            # shape (batch_size), filled with the sequence length size of the sequence_tensor.
+            sequence_lengths = ones_like(sequence_tensor[:, 0, 0]).long() * sequence_tensor.size(1)
+
+        # shape (batch_size, num_spans)
+        end_sentinel_mask = (exclusive_span_ends == sequence_lengths.unsqueeze(-1)).long()
+
+        # Forward Direction: start indices are exclusive. Shape (batch_size, num_spans, input_size / 2)
+        forward_start_embeddings = batched_index_select(forward_sequence, exclusive_span_starts)
+        # Forward Direction: end indices are inclusive, so we can just use span_ends.
+        # Shape (batch_size, num_spans, input_size / 2)
+        forward_end_embeddings = batched_index_select(forward_sequence, span_ends)
+
+        # Backward Direction: start indices are actually the end indices, because we
+        # are going backwards, which we want to be inclusive, so we can just use span_starts.
+        # Shape (batch_size, num_spans, input_size / 2)
+        backward_start_embeddings = batched_index_select(backward_sequence, span_starts)
```

Comment:
I think this would be less confusing if you called this `backward_end_embeddings`.